src/buffer.c

```diff
@@ -39,11 +40,14 @@
 #define MIN(a, b) ((a) < (b) ? (a) : (b))
 #define NOT_POWER_OF_2(x) (x == 0 || (x & (x - 1)))
 
+/* How much extra data to store for SOCK_DGRAM reads */
+#define EXTRA_SOCK_DGRAM 2
+
 
 static const size_t BUFFER_MAX_SIZE = 1024 * 1024 * 1024;
 
 
-static size_t setup_write_iov(const struct Buffer *, struct iovec *, size_t);
+static size_t setup_write_iov(const struct Buffer *, struct iovec *, size_t, size_t *);
```

Comment:
Oh, this is looking complicated how about this instead: SOCK_DGRAM buffers are always 16-bit aligned. 

```
static uint16_t *
reserve_dgram_length(struct Buffer *buffer) {

    if (buffer_size(dst) - dst->len < sizeof(uint16_t) + 1)
        return NULL; /* insufficient room */

    /* Reserve remove for dgram length prefix */
    uint16_t *dgram_len = (uint16_t *)&buffer->buffer[
        (buffer->head + buffer->len) & buffer->size_mask];

    buffer->len += sizeof(uint16_t);

    return dgram_lenn;
}

#define ALIGN_16BIT(x) ((x + 1) & ~((typeof(x))1))

/*
 * Read data from file into buffer
 */
ssize_t
buffer_read(struct Buffer *buffer, int fd) {
    /* coalesce when reading into an empty buffer */
    if (buffer->len == 0)
        buffer->head = 0;

    struct iovec iov[2];
    ssize_t bytes;

    if (buffer->type == SOCK_DGRAM) {
        uint16_t *dgram_len = reserve_dgram_length(buffer);

        size_t iov_len = setup_write_iov(buffer, iov, 0);
        bytes = readv(fd, iov, iov_len);

        if (bytes < 0)
            /* Release dgram length prefix space */
            buffer->len -= sizeof(uint16_t);
        else
            /* Store the dgram length */
            *dgram_len = (uint16_t)bytes;

        if (bytes > 0)
            advance_write_position(buffer, (size_t)ALIGN_16BIT(bytes));
    } else {
        size_t iov_len = setup_write_iov(buffer, iov, 0);
        bytes = readv(fd, iov, iov_len);

        if (bytes > 0)
            advance_write_position(buffer, (size_t)bytes);
    }

    return bytes;
}

/*
 * Write data to file from buffer
 */
ssize_t
buffer_write(struct Buffer *buffer, int fd) {
    struct iovec iov[2];
    ssize_t bytes;

    if (buffer->type == SOCK_DGRAM) {
        uint16_t *dgram_len = (uint16_t *)&buffer->buffer[buffer->head];
        buffer->head = (buffer->head + sizeof(uint16_t)) & buffer->size_mask;
        buffer->len -= sizeof(uint16_t);

        size_t iov_len = setup_read_iov(buffer, iov, *dgram_len);
        ssize_t bytes = writev(fd, iov, iov_len);

        if (bytes < 0) {
            /* restore dgram length prefix */
            buffer->head = (buffer->head - sizeof(uint16_t)) & buffer->size_mask;
            buffer->len += sizeof(uint16_t);
        }

        if (bytes > 0)
            advance_read_position(buffer, (size_t)ALIGN_16BIT(bytes));
    } else {
        size_t iov_len = setup_read_iov(buffer, iov, 0);
        ssize_t bytes = writev(fd, iov, iov_len);

        if (bytes > 0)
            advance_read_position(buffer, (size_t)bytes);
    }

    return bytes;
}
```

Reply:
Much better! I'm taking this example and reworking it now into the rest of the buffer code, thanks!